devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
+
+- `get` requests
+- `pop` requests
```

Comment:
Is "get" like "peek"? Do we actually need "get", or would "pop" alone work?

Reply:
"get"  is a little like "peek", but "peek" only allows you to see the head of the queue whereas "get" allows you to iterate over the entire queue.

In GeoReplication terms, each `BeginReplicate` operation needs be gotten (by the secondary state provider calling `get_next`) and acknowledged. So this would require two passes over each job(each "job" is a `BeginReplicate` call by the primary state provider).

Now that makes me think... What if instead of having two passes over each job, we used to `job_queue`s. One for serving `get_next` the other for serving `acknowledge`. 

Hmm...